File: Customer Data Analysis and Insights Project/scripts/python/ml_analysis.py
"""
Machine Learning Analysis
Customer Data Analysis - Customer Segmentation
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.cluster import KMeans, AgglomerativeClustering, DBSCAN
from sklearn.metrics import silhouette_score, davies_bouldin_score, calinski_harabasz_score
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
import warnings
from pathlib import Path

warnings.filterwarnings('ignore')

# Set style
sns.set_style('whitegrid')
plt.rcParams['figure.figsize'] = (14, 8)

# Set paths
data_path = Path('../../data/Customers.csv')
results_path = Path('../../results')
results_path.mkdir(exist_ok=True)

# Load data
df = pd.read_csv(data_path)

def prepare_data():
    """Prepare data for ML analysis"""
    print("=" * 50)
    print("DATA PREPARATION")
    print("=" * 50)
    
    # Encode categorical variables
    le_country = LabelEncoder()
    le_city = LabelEncoder()
    
    df_encoded = df.copy()
    df_encoded['Country_encoded'] = le_country.fit_transform(df['Country'])
    df_encoded['City_encoded'] = le_city.fit_transform(df['City'])
    
    # Select features
    features = ['CustomerID', 'Country_encoded', 'City_encoded']
    X = df_encoded[features].values
    
    # Standardize features
    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(X)
    
    print(f"Data shape: {X_scaled.shape}")
    print(f"Features: {features}")
    
    return X_scaled, df_encoded, le_country, le_city, features

def determine_optimal_clusters(X_scaled, max_clusters=10):
    """Determine optimal number of clusters using Elbow method and Silhouette score"""
    print("\n" + "=" * 50)
    print("DETERMINING OPTIMAL NUMBER OF CLUSTERS")
    print("=" * 50)
    
    inertias = []
    silhouette_scores = []
    K_range = range(2, min(max_clusters + 1, len(X_scaled)))
    
    for k in K_range:
        kmeans = KMeans(n_clusters=k, random_state=42, n_init=10)
        kmeans.fit(X_scaled)
        inertias.append(kmeans.inertia_)
        silhouette_scores.append(silhouette_score(X_scaled, kmeans.labels_))
    
    # Plot Elbow method
    fig, axes = plt.subplots(1, 2, figsize=(16, 6))
    
    axes[0].plot(K_range, inertias, 'bo-', linewidth=2, markersize=8)
    axes[0].set_xlabel('Number of Clusters (k)')
    axes[0].set_ylabel('Inertia')
    axes[0].set_title('Elbow Method for Optimal k')
    axes[0].grid(True, alpha=0.3)
    
    axes[1].plot(K_range, silhouette_scores, 'ro-', linewidth=2, markersize=8)
    axes[1].set_xlabel('Number of Clusters (k)')
    axes[1].set_ylabel('Silhouette Score')
    axes[1].set_title('Silhouette Score for Optimal k')
    axes[1].grid(True, alpha=0.3)
    
    plt.tight_layout()
    plt.savefig(results_path / 'optimal_clusters.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # Find optimal k (highest silhouette score)
    optimal_k = K_range[np.argmax(silhouette_scores)]
    print(f"\nOptimal number of clusters: {optimal_k}")
    print(f"Silhouette score: {max(silhouette_scores):.4f}")
    
    return optimal_k

def kmeans_clustering(X_scaled, n_clusters, df_encoded):
    """Perform K-Means clustering"""
    print("\n" + "=" * 50)
    print("K-MEANS CLUSTERING")
    print("=" * 50)
    
    # Fit K-Means
    kmeans = KMeans(n_clusters=n_clusters, random_state=42, n_init=10)
    clusters = kmeans.fit_predict(X_scaled)
    
    # Add cluster labels to dataframe
    df_clustered = df_encoded.copy()
    df_clustered['Cluster'] = clusters
    
    # Evaluate clustering
    silhouette = silhouette_score(X_scaled, clusters)
    davies_bouldin = davies_bouldin_score(X_scaled, clusters)
    calinski_harabasz = calinski_harabasz_score(X_scaled, clusters)
    
    print(f"Silhouette Score: {silhouette:.4f}")
    print(f"Davies-Bouldin Score: {davies_bouldin:.4f}")
    print(f"Calinski-Harabasz Score: {calinski_harabasz:.4f}")
    
    # Cluster statistics
    print("\n=== Cluster Statistics ===")
    cluster_stats = df_clustered.groupby('Cluster').agg({
        'CustomerID': 'count',
        'Country_encoded': 'mean',
        'City_encoded': 'mean'
    }).round(2)
    cluster_stats.columns = ['Customer_Count', 'Avg_Country_encoded', 'Avg_City_encoded']
    print(cluster_stats)
    
    # Country distribution by cluster
    print("\n=== Country Distribution by Cluster ===")
    country_cluster = pd.crosstab(df_clustered['Cluster'], df_clustered['Country'])
    print(country_cluster)
    
    # Visualize clusters using PCA
    pca = PCA(n_components=2)
    X_pca = pca.fit_transform(X_scaled)
    
    plt.figure(figsize=(12, 8))
    scatter = plt.scatter(X_pca[:, 0], X_pca[:, 1], c=clusters, cmap='viridis', alpha=0.6, s=100)
    plt.scatter(kmeans.cluster_centers_[:, 0], kmeans.cluster_centers_[:, 1], 
                c='red', marker='x', s=200, linewidths=3, label='Centroids')
    plt.xlabel(f'PC1 ({pca.explained_variance_ratio_[0]:.2%} variance)')
    plt.ylabel(f'PC2 ({pca.explained_variance_ratio_[1]:.2%} variance)')
    plt.title(f'K-Means Clustering (k={n_clusters})')
    plt.colorbar(scatter, label='Cluster')
    plt.legend()
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig(results_path / 'kmeans_clustering.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    return df_clustered, kmeans

def hierarchical_clustering(X_scaled, n_clusters, df_encoded):
    """Perform Hierarchical Clustering"""
    print("\n" + "=" * 50)
    print("HIERARCHICAL CLUSTERING")
    print("=" * 50)
    
    # Fit Hierarchical Clustering
    hierarchical = AgglomerativeClustering(n_clusters=n_clusters, linkage='ward')
    clusters = hierarchical.fit_predict(X_scaled)
    
    # Add cluster labels
    df_hierarchical = df_encoded.copy()
    df_hierarchical['Cluster'] = clusters
    
    # Evaluate clustering
    silhouette = silhouette_score(X_scaled, clusters)
    davies_bouldin = davies_bouldin_score(X_scaled, clusters)
    calinski_harabasz = calinski_harabasz_score(X_scaled, clusters)
    
    print(f"Silhouette Score: {silhouette:.4f}")
    print(f"Davies-Bouldin Score: {davies_bouldin:.4f}")
    print(f"Calinski-Harabasz Score: {calinski_harabasz:.4f}")
    
    # Visualize clusters using PCA
    pca = PCA(n_components=2)
    X_pca = pca.fit_transform(X_scaled)
    
    plt.figure(figsize=(12, 8))
    plt.scatter(X_pca[:, 0], X_pca[:, 1], c=clusters, cmap='plasma', alpha=0.6, s=100)
    plt.xlabel(f'PC1 ({pca.explained_variance_ratio_[0]:.2%} variance)')
    plt.ylabel(f'PC2 ({pca.explained_variance_ratio_[1]:.2%} variance)')
    plt.title(f'Hierarchical Clustering (k={n_clusters})')
    plt.colorbar(label='Cluster')
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig(results_path / 'hierarchical_clustering.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    return df_hierarchical

def cluster_analysis(df_clustered):
    """Analyze cluster characteristics"""
    print("\n" + "=" * 50)
    print("CLUSTER ANALYSIS")
    print("=" * 50)
    
    # Country distribution by cluster
    print("\n=== Top Countries by Cluster ===")
    for cluster_id in sorted(df_clustered['Cluster'].unique()):
        cluster_data = df_clustered[df_clustered['Cluster'] == cluster_id]
        top_countries = cluster_data['Country'].value_counts().head(5)
        print(f"\nCluster {cluster_id} ({len(cluster_data)} customers):")
        for country, count in top_countries.items():
            print(f"  {country}: {count} customers ({(count/len(cluster_data)*100):.1f}%)")
    
    # City distribution by cluster
    print("\n=== Top Cities by Cluster ===")
    for cluster_id in sorted(df_clustered['Cluster'].unique()):
        cluster_data = df_clustered[df_clustered['Cluster'] == cluster_id]
        top_cities = cluster_data['City'].value_counts().head(5)
        print(f"\nCluster {cluster_id}:")
        for city, count in top_cities.items():
            print(f"  {city}: {count} customers ({(count/len(cluster_data)*100):.1f}%)")
    
    # Visualize cluster composition
    fig, axes = plt.subplots(1, 2, figsize=(16, 6))
    
    # Country distribution
    country_cluster = pd.crosstab(df_clustered['Cluster'], df_clustered['Country'])
    country_cluster.plot(kind='bar', ax=axes[0], stacked=True, colormap='tab20')
    axes[0].set_xlabel('Cluster')
    axes[0].set_ylabel('Number of Customers')
    axes[0].set_title('Country Distribution by Cluster')
    axes[0].legend(bbox_to_anchor=(1.05, 1), loc='upper left')
    axes[0].grid(axis='y', alpha=0.3)
    
    # Cluster sizes
    cluster_sizes = df_clustered['Cluster'].value_counts().sort_index()
    cluster_sizes.plot(kind='bar', ax=axes[1], color='steelblue', edgecolor='black')
    axes[1].set_xlabel('Cluster')
    axes[1].set_ylabel('Number of Customers')
    axes[1].set_title('Cluster Sizes')
    axes[1].grid(axis='y', alpha=0.3)
    
    plt.tight_layout()
    plt.savefig(results_path / 'cluster_analysis.png', dpi=300, bbox_inches='tight')
    plt.close()

def main():
    """Main function"""
    # Prepare data
    X_scaled, df_encoded, le_country, le_city, features = prepare_data()
    
    # Determine optimal number of clusters
    optimal_k = determine_optimal_clusters(X_scaled, max_clusters=min(8, len(X_scaled)//2))
    
    # Perform K-Means clustering
    df_clustered, kmeans = kmeans_clustering(X_scaled, optimal_k, df_encoded)
    
    # Perform Hierarchical clustering
    df_hierarchical = hierarchical_clustering(X_scaled, optimal_k, df_encoded)
    
    # Cluster analysis
    cluster_analysis(df_clustered)
    
    # Save results
    df_clustered.to_csv(results_path / 'clustered_customers.csv', index=False)
    print(f"\n✓ Clustered data saved to {results_path / 'clustered_customers.csv'}")
    
    print("\n" + "=" * 50)
    print("ML ANALYSIS SUMMARY")
    print("=" * 50)
    print(f"\n1. Optimal number of clusters: {optimal_k}")
    print("2. Performed K-Means clustering")
    print("3. Performed Hierarchical clustering")
    print("4. Analyzed cluster characteristics")
    print("5. Generated visualizations and saved results")

if __name__ == "__main__":
    main()

File: Customer Data Analysis and Insights Project/scripts/r/ml_analysis.R
# Machine Learning Analysis
# Customer Data Analysis - Customer Segmentation

# Load libraries
library(tidyverse)
library(ggplot2)
library(dplyr)
library(cluster)
library(factoextra)
library(NbClust)
library(corrplot)

# Set paths
data_path <- "../../data/Customers.csv"
results_path <- "../../results"

# Create results directory if it doesn't exist
if (!dir.exists(results_path)) {
  dir.create(results_path, recursive = TRUE)
}

# Load data
df <- read.csv(data_path, stringsAsFactors = FALSE)

cat("==================================================\n")
cat("MACHINE LEARNING ANALYSIS\n")
cat("==================================================\n\n")

# 1. Data Preparation
cat("=== DATA PREPARATION ===\n")

# Encode categorical variables
df_encoded <- df
if ("Country" %in% colnames(df)) {
  df_encoded$Country_encoded <- as.numeric(as.factor(df$Country))
}
if ("City" %in% colnames(df)) {
  df_encoded$City_encoded <- as.numeric(as.factor(df$City))
}

# Select features
features <- c("CustomerID")
if ("Country_encoded" %in% colnames(df_encoded)) {
  features <- c(features, "Country_encoded")
}
if ("City_encoded" %in% colnames(df_encoded)) {
  features <- c(features, "City_encoded")
}

X <- df_encoded[, features, drop = FALSE]

# Standardize features
X_scaled <- scale(X)

cat("Data shape:", dim(X_scaled), "\n")
cat("Features:", paste(features, collapse = ", "), "\n")

# 2. Determine Optimal Number of Clusters
cat("\n=== DETERMINING OPTIMAL NUMBER OF CLUSTERS ===\n")

# Elbow method and Silhouette score
max_clusters <- min(8, nrow(X_scaled) %/% 2)
K_range <- 2:max_clusters

inertias <- numeric(length(K_range))
silhouette_scores <- numeric(length(K_range))

for (i in seq_along(K_range)) {
  k <- K_range[i]
  kmeans_result <- kmeans(X_scaled, centers = k, nstart = 10)
  inertias[i] <- kmeans_result$tot.withinss
  silhouette_scores[i] <- mean(silhouette(kmeans_result$cluster, dist(X_scaled))[, 3])
}

# Plot Elbow method and Silhouette score
png(file = file.path(results_path, "optimal_clusters.png"), 
    width = 1600, height = 600, res = 300)
par(mfrow = c(1, 2))
plot(K_range, inertias, type = "b", pch = 19, col = "steelblue",
     xlab = "Number of Clusters (k)", ylab = "Inertia",
     main = "Elbow Method for Optimal k")
plot(K_range, silhouette_scores, type = "b", pch = 19, col = "coral",
     xlab = "Number of Clusters (k)", ylab = "Silhouette Score",
     main = "Silhouette Score for Optimal k")
dev.off()

# Find optimal k
optimal_k <- K_range[which.max(silhouette_scores)]
cat("Optimal number of clusters:", optimal_k, "\n")
cat("Silhouette score:", max(silhouette_scores), "\n")

# 3. K-Means Clustering
cat("\n=== K-MEANS CLUSTERING ===\n")

kmeans_result <- kmeans(X_scaled, centers = optimal_k, nstart = 10)
clusters <- kmeans_result$cluster

# Add cluster labels to dataframe
df_clustered <- df_encoded
df_clustered$Cluster <- clusters

# Evaluate clustering
silhouette_score <- mean(silhouette(clusters, dist(X_scaled))[, 3])
cat("Silhouette Score:", silhouette_score, "\n")

# Cluster statistics
cat("\n=== Cluster Statistics ===\n")
cluster_stats <- df_clustered %>%
  group_by(Cluster) %>%
  summarise(
    Customer_Count = n(),
    Avg_Country_encoded = mean(Country_encoded, na.rm = TRUE),
    Avg_City_encoded = mean(City_encoded, na.rm = TRUE),
    .groups = "drop"
  )
print(cluster_stats)

# Country distribution by cluster
if ("Country" %in% colnames(df_clustered)) {
  cat("\n=== Country Distribution by Cluster ===\n")
  country_cluster <- table(df_clustered$Cluster, df_clustered$Country)
  print(country_cluster)
}

# Visualize clusters using PCA
pca_result <- prcomp(X_scaled)
X_pca <- pca_result$x[, 1:2]

png(file = file.path(results_path, "kmeans_clustering.png"), 
    width = 1200, height = 800, res = 300)
plot(X_pca, col = clusters, pch = 19, 
     main = paste("K-Means Clustering (k =", optimal_k, ")"),
     xlab = paste("PC1 (", round(summary(pca_result)$importance[2, 1] * 100, 2), "% variance)", sep = ""),
     ylab = paste("PC2 (", round(summary(pca_result)$importance[2, 2] * 100, 2), "% variance)", sep = ""))
points(kmeans_result$centers[, 1:2], col = "red", pch = 4, cex = 2, lwd = 3)
legend("topright", legend = paste("Cluster", 1:optimal_k), 
       col = 1:optimal_k, pch = 19, title = "Clusters")
dev.off()

# 4. Hierarchical Clustering
cat("\n=== HIERARCHICAL CLUSTERING ===\n")

# Compute distance matrix
dist_matrix <- dist(X_scaled, method = "euclidean")

# Perform hierarchical clustering
hierarchical_result <- hclust(dist_matrix, method = "ward.D2")
clusters_hierarchical <- cutree(hierarchical_result, k = optimal_k)

# Add cluster labels
df_hierarchical <- df_encoded
df_hierarchical$Cluster <- clusters_hierarchical

# Evaluate clustering
silhouette_score_hierarchical <- mean(silhouette(clusters_hierarchical, dist_matrix)[, 3])
cat("Silhouette Score:", silhouette_score_hierarchical, "\n")

# Visualize dendrogram
png(file = file.path(results_path, "hierarchical_dendrogram.png"), 
    width = 1400, height = 800, res = 300)
plot(hierarchical_result, main = "Hierarchical Clustering Dendrogram",
     xlab = "Customers", sub = "", cex = 0.6)
rect.hclust(hierarchical_result, k = optimal_k, border = "red")
dev.off()

# Visualize clusters
png(file = file.path(results_path, "hierarchical_clustering.png"), 
    width = 1200, height = 800, res = 300)
plot(X_pca, col = clusters_hierarchical, pch = 19,
     main = paste("Hierarchical Clustering (k =", optimal_k, ")"),
     xlab = paste("PC1 (", round(summary(pca_result)$importance[2, 1] * 100, 2), "% variance)", sep = ""),
     ylab = paste("PC2 (", round(summary(pca_result)$importance[2, 2] * 100, 2), "% variance)", sep = ""))
legend("topright", legend = paste("Cluster", 1:optimal_k), 
       col = 1:optimal_k, pch = 19, title = "Clusters")
dev.off()

# 5. Cluster Analysis
cat("\n=== CLUSTER ANALYSIS ===\n")

# Country distribution by cluster
if ("Country" %in% colnames(df_clustered)) {
  cat("\n=== Top Countries by Cluster ===\n")
  for (cluster_id in sort(unique(df_clustered$Cluster))) {
    cluster_data <- df_clustered[df_clustered$Cluster == cluster_id, ]
    top_countries <- cluster_data %>%
      count(Country) %>%
      arrange(desc(n)) %>%
      head(5)
    
    cat("\nCluster", cluster_id, "(", nrow(cluster_data), "customers):\n", sep = "")
    for (i in 1:nrow(top_countries)) {
      country <- top_countries$Country[i]
      count <- top_countries$n[i]
      percentage <- (count / nrow(cluster_data)) * 100
      cat("  ", country, ": ", count, " customers (", round(percentage, 1), "%)\n", sep = "")
    }
  }
}

# City distribution by cluster
if ("City" %in% colnames(df_clustered)) {
  cat("\n=== Top Cities by Cluster ===\n")
  for (cluster_id in sort(unique(df_clustered$Cluster))) {
    cluster_data <- df_clustered[df_clustered$Cluster == cluster_id, ]
    top_cities <- cluster_data %>%
      count(City) %>%
      arrange(desc(n)) %>%
      head(5)
    
    cat("\nCluster", cluster_id, ":\n", sep = "")
    for (i in 1:nrow(top_cities)) {
      city <- top_cities$City[i]
      count <- top_cities$n[i]
      percentage <- (count / nrow(cluster_data)) * 100
      cat("  ", city, ": ", count, " customers (", round(percentage, 1), "%)\n", sep = "")
    }
  }
}

# Visualize cluster composition
if ("Country" %in% colnames(df_clustered)) {
  png(file = file.path(results_path, "cluster_analysis.png"), 
      width = 1600, height = 600, res = 300)
  par(mfrow = c(1, 2))
  
  # Country distribution
  country_cluster_table <- table(df_clustered$Cluster, df_clustered$Country)
  barplot(country_cluster_table, main = "Country Distribution by Cluster",
          xlab = "Country", ylab = "Number of Customers", 
          legend.text = paste("Cluster", 1:optimal_k),
          args.legend = list(x = "topright", cex = 0.8))
  
  # Cluster sizes
  cluster_sizes <- table(df_clustered$Cluster)
  barplot(cluster_sizes, main = "Cluster Sizes",
          xlab = "Cluster", ylab = "Number of Customers",
          col = "steelblue", border = "black")
  
  dev.off()
}

# Save results
write.csv(df_clustered, file = file.path(results_path, "clustered_customers.csv"), row.names = FALSE)
cat("\n✓ Clustered data saved to", file.path(results_path, "clustered_customers.csv"), "\n")

cat("\n==================================================\n")
cat("ML ANALYSIS SUMMARY\n")
cat("==================================================\n")
cat("\n1. Optimal number of clusters:", optimal_k, "\n")
cat("2. Performed K-Means clustering\n")
cat("3. Performed Hierarchical clustering\n")
cat("4. Analyzed cluster characteristics\n")
cat("5. Generated visualizations and saved results\n")

cat("\n✓ ML analysis completed successfully!\n")

